source/Server/IntegratedAuthenticationModule.cs

```diff
@@ -18,26 +22,34 @@ public IntegratedAuthenticationModule(ILog log, IAuthCookieCreator tokenIssuer,
                     return responseCreator.Unauthorized(Request);
 
                 var principal = (IOctopusPrincipal)Context.CurrentUser;
-                var authCookies = tokenIssuer.CreateAuthCookies(Context.Request, principal.IdentificationToken, SessionExpiry.TwentyDays);
 
-                var whitelist = authenticationConfigurationStore.GetTrustedRedirectUrls();
-                Response response;
-                if (Request.Query["redirectTo"].HasValue && Requests.IsLocalUrl(Request.Query["redirectTo"].Value, whitelist))
+                if (Request.Query["state"].HasValue)
                 {
-                    var redirectLocation = Request.Query["redirectTo"].Value;
-                    response = new RedirectResponse(redirectLocation).WithCookies(authCookies);
-                }
-                else
-                {
-                    if (Request.Query["redirectTo"].HasValue)
+                    var stateData = HttpUtility.UrlDecode(Request.Query["state"].Value);
+                    var state = JsonConvert.DeserializeObject<LoginState>(stateData);
+
+                    var authCookies = tokenIssuer.CreateAuthCookies(Context.Request, principal.IdentificationToken, SessionExpiry.TwentyDays, (bool)state.UsingSecureConnection);
+
+                    var whitelist = authenticationConfigurationStore.GetTrustedRedirectUrls();
+
+                    if (Requests.IsLocalUrl(state.RedirectAfterLoginTo, whitelist))
                     {
-                        log.WarnFormat("Prevented potential Open Redirection attack on an NTLM challenge, to the non-local url {0}", Request.Query["redirectTo"].Value);
+                        return new RedirectResponse(state.RedirectAfterLoginTo).WithCookies(authCookies);
                     }
+                    else
+                    {
+                        if (!string.IsNullOrWhiteSpace(state.RedirectAfterLoginTo))
+                        {
+                            log.WarnFormat(
+                                "Prevented potential Open Redirection attack on an NTLM challenge, to the non-local url {0}",
+                                Request.Query["redirectTo"].Value);
+                        }
 
-                    response = new RedirectResponse(Request.Url.BasePath ?? "/").WithCookies(authCookies);
+                        return new RedirectResponse(Request.Url.BasePath ?? "/").WithCookies(authCookies);
+                    }
                 }
 
-                return response;
+                return responseCreator.BadRequest("Invalid state passed to server");
```

Comment:
I feel like this message could use a bit more context to make it easier to support. Perhaps `Invalid state passed to the server as part of the NTLM challenge.`